docs: v12 migration guide [LIBS-689] (#884)
dhis2/app-platform: docs/migration/v12.md

> # Platform v12 migration guide: Vite & React 18

Reviewer: it's a great guide @KaiVandivier - two comments:

- Shall we have a mention for the changes to better supprot TypeScript?
- Can we already have this merged with a disclaimer that it is still in alpha, so that people can give it a go?

Author: 1. Definitely! Are these the things you're thinking of?
    1. Bootstrap a typescript app with `yarn d2 app scripts init --typescript my-app-name`
    2. TypeScript is supported out of the box with Vite [Is this new, compared to CRA?]
    3. [Upcoming] Use `@dhis2/cli-style` runs type checking
4. Sure, we could merge this to master, but I'm kinda hoping the rest of alpha will be in master soon too haha

Reviewer: TS was also supported in CRA, but I think it's still worth it to mention it's supported in Vite .. we can also say that we had other improvements for types like adding types to UI library, and to-come-soon types for the API generated out of OpenAPI specs

Author: Added some text in the new & upcoming changes sections 👍

Examples come from the migration for the Aggregate Data Entry app; you can check out the PR [here](https://github.com/dhis2/aggregate-data-entry-app/pull/396) for more examples in context.

##### `userEvent`

`userEvent` is async now in `@testing-library/user-event`. Check the [release notes](https://github.com/testing-library/user-event/releases/tag/v14.0.0) for an idea of the improvements and new API.

In practice, this means mostly that you'd need to add an `await` for `userEvent` interactions like `userEvent.click()`:

```diff
+ import userEvent from '@testing-library/user-event'

- it('should allow re-running validation', () => {
+ // Make the function async:
+ it('should allow re-running validation', async () => {
    // ...

-   userEvent.click(getByText('Run validation again'))
-   // Or this other form:
-   fireEvent.click(getByText('Run validation again'))
+   // Now that it's asynchronous, await the event:
+   await userEvent.click(getByText('Run validation again'))

    await findByText('2 medium priority alerts')
    expect(queryByText('There was a problem running validation')).toBeNull()
})

> ```

Author: Cool, added more to this section 👍

4. Flow types won’t work, although this is a motivator for extensible config
5. Import aliases will likely break (also a motivator for extensible config)
6. `global.variableName` is no longer supported; use `window.variableName` instead
7. In order to make the dev server available on LAN, the `--host` flag needs to be passed to the start script
8. Workers imported in the “webpack” way will need to be imported in a new way: [docs](https://vitejs.dev/guide/features.html#web-workers)
9. App.jsx entrypoint is no longer quietly added if a plugin entrypoint is defined or for a defined but empty entrypoints object
10. Plugins and apps are started together in the same process on the same port
11. The `PORT` env var is removed
12. `import * as MyClass from ‘my-module’` is now more restrictive, and may break in some cases. So far, changing the import to `import MyClass from ‘my-module’` has fixed it. DOMPurify is one example case, noticed in the Login App

### Deprecations

-   Environment variables on `process.env` will be dropped in future versions, in favor of variables on `import.meta.env`
-   The default `direction` config value is `ltr`; in the future, it will be `auto`

> -   `--allowJsxInJs` will be removed

Reviewer: wil this be removed in the underlying Vite CLI?

Author: Not sure which bullet point you mean haha, but here's an answer for each:
1. Env vars are under our control, but I'm actually rethinking this point, i.e. maybe we should keep variables on `process.env` -- Earlier, I thought `import.meta.env` was a pretty standard thing with ES6 and that we should migrate to that for env vars, but I think that's actually just a Vite pattern, so it might be "more standard" to keep env vars on `process.env`. If we do that, we can at least deprecate the `REACT_APP_` prefixes for the variable keys
2. (Probably obvious, but direction is also under our control)
3. For `--allowJsxInJs`, that's also under our control -- it's a flag that our own CLI reads, and adds a bunch of kinda complex Vite config in `makeViteConfig.mjs`. That additional config may or may not be supported in future versions of Vite, but I think we should remove _our_ `--allowJsxInJs` flag and config in the next `cli-app-scripts` version to make things simpler, standardized, and more performant

> # Platform v12 migration guide: Vite & React 18

Author: Added in [intro revisions](https://github.com/dhis2/app-platform/pull/884/commits/071f5dc8e22aafd60318ba69ac353da608326cd2), but added v12 since it's helpful to see the relevant package version

> Good news! Vite and React 18 in the app platform are ready to use!

Author: Moved this kind of language into the dev portal blog post: https://github.com/dhis2/developer-portal/pull/167

We’re very excited for these updates — both will significantly modernize the App Platform, and Vite will be a big upgrade from Create React App, which we used under the hood to build and serve apps previously. It will greatly improve the developer experience, and with greater control over the configuration, it will open up some powerful new possibilities for the platform.

Here are some tips about what to expect, and how to easily upgrade to the latest version of `@dhis2/cli-app-scripts` to take advantage of Vite and React 18.

### Notable changes

> These are some things that you’ll see right away:

Reviewer 2: ```suggestion
These are some things that you’ll see right away after upgrading:
```

> -   It’s fast! Starting up an app is nearly instant, and building an app is about 3-4 times faster than before

Reviewer 2: ```suggestion
-   It’s fast! Starting up an app is nearly instant, and building an app is about 3-4 times faster compared to Create React App (CRA)
```

-   Plugins are handled better:
    -   Start-up of both the app and plugin is nearly instant
    -   The app and plugin are run on the same port
    -   Support for a plugin without an app is improved
    -   Code is shared between entrypoints, which makes bundles smaller

>     -   HMR for code changes in the plugin will be as fast as in the app

Reviewer 2: ```suggestion
    -   Hot Module Replacement (HMR) for code changes in the plugin will be as fast as in the app
```

-   There are some new TypeScript features:

>     -   Bootstrapping an app with a TypeScript template is supported: `d2-app-scripts init --typescript my-app`

Reviewer 2: Is this command explained elsewhere? If so, can we link it? (or document it elsewhere if not)

    -   Vite has native support for TypeScript
    -   Although not in the App Platform package, with the latest `@dhis2/cli-style`, TypeScript type checking is performed when running `d2-style lint`
    -   With `@dhis2/ui` version 9, the UI library is now typed as well
-   There’s a small suite of tools available in the CLI when running an app in dev mode:

>     -   With the dev server running, press `h + enter` to see the options

Reviewer 2: Can you link the vite docs about these tools?

Reviewer 2: Oh wow really? That's why I couldn't find it I guess

    -   Options include exposing the server on LAN, opening the app in the browser, cleanly quitting the server, and restarting the dev server (which can be helpful if you’re modifying libraries in node_modules — see more below)
-   The build output includes a summary to inspect chunks

### Future changes

> With Vite, the door is open for some big future improvements:

Reviewer 2: Just a note about this section, this will need adjusted once these features are either deployed or dismissed. I think it is fine to have it here but we must not forget. 

If there are already JIRA tickets/epics about these features, we should link it to this file as a todo.

### Allow JSX in JS temporarily

To make it as fast as possible to upgrade and get up and running with Vite and React 18, we added an `--allowJsxInJs` flag that you can pass to `start` and `build` scripts, so that you can test out the changes without any file renaming. In many cases, you should be able to upgrade your platform dependencies and run `start` with the flag and see your app running! Other times, you may need to add a few tweaks — see the troubleshooting and breaking changes sections below.

Due to the downsides mentioned above, however, we will remove the `--allowJsxInJs` option in the next major version. When you are ready to rename your files, you can use the migration script we’ve made.

### .js to .jsx migration script

We made a codemod that can migrate your codebase to the right file extensions quickly! It checks .js files for JSX syntax, updates the extension if needed, then updates imports within the file tree to target the new filenames.

```sh
yarn d2-app-scripts migrate js-to-jsx
```

Check out the [script docs](../scripts/migrate/js-to-jsx.md) for options, tips, and caveats.

## React 18

> In terms of the React 18-related changes, upgrading to `@dhis2/cli-app-scripts` version 12 should generally a smooth process for the consuming apps. Most of the [changes](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) shouldn't affect most apps.

Reviewer 2: I feel wording could be changed slightly here for readability. Made a suggestion

```suggestion
With this update we also upgraded from React 16 to React 18. Upgrading to `@dhis2/cli-app-scripts` version 12 should generally a smooth process for the consuming apps. Most of the [changes](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) shouldn't affect most apps.
```

### Default props

The main change in React 18 that is expected to affect apps at runtime is that using `MyComponent.defaultProps = { ... }` is deprecated, and will be removed in the next major version. While non-breaking, this change comes along with warning messages in the console for every rendered component that uses them, which can be chatty and annoying.

The fix is to move prop defaults -- either to default parameters for functional components, or to a static property for class components. This code mod can help with the migration, and was used for the `@dhis2/ui` library: [transform.js](https://gist.github.com/kabaros/45ed16660ac89dc48d22e03836b1e981)

> **NB!** Note that default params for functional components are computed on each render, so default values like arrays and objects as defaults may unintentionally retrigger `useEffect` and `useMemo` functions on each render. Instead, use a stable reference like so:

Reviewer 2: Might be nice to use an admonition here.

```suggestion
:::info Note
Default params for functional components are computed on each render, so default values like arrays and objects as defaults may unintentionally retrigger `useEffect` and `useMemo` functions on each render. Instead, use a stable reference like so:
```

Should also add `:::` below the codeblock

```js
const arr = []
const MyComponent = ({ options = arr }) => {
    /* ... */
}
```

> 

Reviewer 2: ```suggestion
:::

```

#### React Testing Library

Testing-library supports React 18 since v13, so the first step is to upgrade that library and the supporting libraries around it. These major versions bumps come with a few breaking changes that we will document here with some examples to make migration easier.

Examples come from the migration for the Aggregate Data Entry app; you can check out the PR [here](https://github.com/dhis2/aggregate-data-entry-app/pull/396) for more examples in context.

##### `userEvent`

`userEvent` is async now in `@testing-library/user-event`. Check the [release notes](https://github.com/testing-library/user-event/releases/tag/v14.0.0) for an idea of the improvements and new API. In practice, this means mostly that you'd need to add an `await` for `userEvent` interactions like `userEvent.click()`.

Actions using the `fireEvent` API should also be migrated to `userEvent`, e.g. `fireEvent.change(selector(), { target: { value: 'input text' } })` should be changed to `userEvent.type(selector(), 'input text')`.

Here are some examples that illustrate the changes:

> ```diff

Reviewer 2: we'll need to do a little testing to see if this will render properly on the developer portal

Author: Ah, it doesn't seem to be working :( I copied this into the blog post PR to test:

![Screenshot 2024-11-04 at 14 50 09](https://github.com/user-attachments/assets/1a39ff1a-a9b9-48ec-ab0a-f8fb6699b869)

I'll look into docusaurus support for something similar

Author: Ah, just needed to do this: https://github.com/dhis2/developer-portal/pull/167/commits/45c55d803b610be22fe30768662ad5d53b5cbcca

![Screenshot 2024-11-04 at 16 19 48](https://github.com/user-attachments/assets/ab835884-6e95-4522-9707-bb2c12456a49)

    // ...

-   userEvent.click(getByText('Run validation again'))
-   // Or this other form:
-   fireEvent.click(getByText('Run validation again'))
+   // Now that it's asynchronous, await the event:
+   await userEvent.click(getByText('Run validation again'))

    await findByText('2 medium priority alerts')
    expect(queryByText('There was a problem running validation')).toBeNull()
})
```

```diff
- const login = async ({ user }) => {
+ // Don't need `user`; will use `userEvent` later:
+ const login = async () => {
-   // Remove fireEvent calls:
-   fireEvent.change(screen.getByLabelText('Username'), {
-       target: { value: 'Fl@klypa.no' },
-   })
-   fireEvent.change(screen.getByLabelText('Password'), {
-       target: { value: 'SolanOgLudvig' },
-   })
+   // Replace with userEvent.type():
+   await userEvent.type(screen.getByLabelText('Username'), 'Fl@klypa.no')
+   await userEvent.type(screen.getByLabelText('Password'), 'SolanOgLudvig')

-   await user.click(screen.getByRole('button', { name: /log in/i }))
+   // Replace `user` with `userEvent`:
+   await userEvent.click(screen.getByRole('button', { name: /log in/i }))
}
```

##### `renderHook`

> `renderHook` is not a separate library anymore; it is part of the `@testing-library/react`, and it has [quite a different API](https://testing-library.com/docs/react-testing-library/api/#renderhook). (NB: Some docs online still point to the old version, which can be confusing.)

Reviewer 2: ```suggestion
`renderHook` is not a separate library anymore; it is part of the `@testing-library/react`, and it has [quite a different API](https://testing-library.com/docs/react-testing-library/api/#renderhook). 

:::warning Note
Some docs online still point to the old version, which can be confusing.
:::
```

+           <Comment item={item} />
+       )

-       userEvent.click(getByRole('button', { name: 'Edit comment' }))
+       const editButton = await findByRole('button', { name: 'Edit comment' })
+       // Set up user event with fake timers:
+       const user = userEvent.setup({
+           advanceTimers: jest.advanceTimersByTime,
+       })
+       await user.click(editButton)

        // ...

        expect(queryByRole('progressbar')).not.toBeInTheDocument()
-       userEvent.click(getByRole('button', { name: 'Save comment' }))
-       expect(getByRole('progressbar')).toBeInTheDocument()
+       // Use the setup from above:
+       const btnSaveComment = await findByRole('button', {
+           name: 'Save comment',
+       })
+       await user.click(btnSaveComment)
+       await findByRole('progressbar')
    })
})
```

#### Enzyme

Enzyme is [considered dead](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl): there is no official adapter for React 18 (or 17), so it's best to [migrate to React Testing Library](https://testing-library.com/docs/react-testing-library/migrate-from-enzyme/) for tests, if possible.

> If it's not currently possible to migrate, there is an [unofficial React 18 adapter](https://github.com/cfaester/enzyme-adapter-react-18) that can help in the mean time.

Reviewer 2: ```suggestion
:::tip
If it's not currently possible to migrate, there is an [unofficial React 18 adapter](https://github.com/cfaester/enzyme-adapter-react-18) that can help in the mean time. However, it is highly recommended this is a temporary solution and you should still schedule migrating away from Enzyme.
:::
```

## Environment variables

### `REACT_APP_` prefix no longer needed

> Now that we're moving away from Create React App, the formatting for environment variable names has gotten simpler.

Reviewer 2: As users are in the process of migrating, this wording seems better.
```suggestion
With the migration away from Create React App, the formatting for environment variable names has gotten simpler.
```

As before, environment variables with names that start with `DHIS2_` are added to the app, and can be picked up from `.env` files or on the command line for example. Before, the platform had to add `REACT_APP_` before `DHIS2_` so that Create React App would add the variable to the app, so a variable like `DHIS2_MY_VAR` in the environment would become accessible on `process.env.REACT_APP_DHIS2_MY_VAR` in the app. Now, the `REACT_APP_` prefix isn't needed, so the variable can be accessed on just `process.env.DHIS2_MY_VAR`.

> The `REACT_APP_`-prefixed variables are now deprecated, but they will be kept around alongside the shorter `DHIS2_` variables in this version for backwards compatibility. For example, both `process.env.REACT_APP_DHIS2_MY_VAR` and `process.env.DHIS2_MY_VAR` will be available. The `REACT_APP_` prefixed variables will be removed in the next version though, so make sure to migrate to just the shorter `DHIS2_` variable names.

Reviewer 2: ```suggestion
The `REACT_APP_`-prefixed variables are now deprecated and will be removed in the next major version. However they will be kept around alongside the shorter `DHIS2_` variables in this version for backwards compatibility. 

For example, both `process.env.REACT_APP_DHIS2_MY_VAR` and `process.env.DHIS2_MY_VAR` will be available. 

:::tip
The `REACT_APP_` prefixed variables will be removed in the next version, so make sure to migrate to the shorter `DHIS2_` variable names.
:::
```

### `import.meta.env`

By default, Vite adds its environment variables to the [`import.meta.env` object](https://vite.dev/guide/env-and-mode#env-variables). The App Platform will add `DHIS2_` env vars to `import.meta.env` as well, and the configured [`envPrefix`](https://vite.dev/config/shared-options.html#envprefix) is `DHIS2_`. Emphasis is given to environment variables on the `process.env` object, however, since it is a more widely-used pattern.

## Troubleshooting & tips

Here are some edge cases that were encountered when upgrading some apps with the new platform version:

> -   There is a bug in Vite: importing from a directory, e.g. `‘./app/‘` where we would assume to get `./app/index.[jt]s`, can incorrectly resolve to a file with a similar name to the directory, e.g. `./App.tsx`. I see this as a bug with Vite, but as a workaround, file and dir names can be changed so they don’t conflict. Noticed in the Maintenance app beta.

Reviewer 2: Can the Maintenance app example be linked/explained?

Author: I think it's probably best just to remove the mention, which I've done -- I think the rest of the comment is enough info. I've also removed similar "noticed in ___" mentions

-   If you’re using `identity-obj-proxy` in your test config, make sure you add it to the project’s dependencies. Previously, some apps got away with using the proxy without an explicitid dependency because the package is a dependency of `react-scripts`

> -   You may see some build warnings coming from the `mathjs` package. Upgrading to a recent version of `mathjs` will handle these warnings. Noticed in analytics app (the analytics library has a dependency on `mathjs`)

Reviewer 2: Does the mention of the Analytics app matter? It might confuse more than solve as the solution is a simple upgrade of a package. Maybe add the command to handle this upgrade instead?

Author: Removed this mention, as described in another comment

## Full list of breaking changes and deprecations

> Larger items are mentioned above, but these lists are included for completeness.

Reviewer 2: ```suggestion
The most likely and impactful breaking changes have been described in detail above. However this doesn't cover all the breaking changes and deprecations introduced in this upgrade. Below is the exhausted list. 
```

### Breaking changes

> 1. Node 18 || 20+ is required

Reviewer 2: ```suggestion
1. Node 18 or Node 20+ is required. Node 19 is not supported.
```

2. JSX in files without .jsx or .tsx extensions is not supported by default
3. React 18: Most significant testing changes are described above; the rest can be read about in the React [migration guide](https://react.dev/blog/2022/03/08/react-18-upgrade-guide)
4. Flow types won’t work, although this is a motivator for extensible config
5. Import aliases will likely break (also a motivator for extensible config)
6. `global.variableName` is no longer supported; use `window.variableName` instead
7. In order to make the dev server available on LAN, the `--host` flag needs to be passed to the start script
8. Workers imported in the “webpack” way will need to be imported in a new way: [docs](https://vitejs.dev/guide/features.html#web-workers)
9. An `App.jsx` entrypoint is no longer quietly added if a plugin entrypoint is defined or for a defined but empty entrypoints object
10. Plugins and apps are started together in the same process on the same port
11. Custom `index.html` files are no longer supported

> 12. `import * as MyClass from ‘my-module’` is now more restrictive, and may break in some cases. So far, changing the import to `import MyClass from ‘my-module’` has fixed it. DOMPurify is one example case, noticed in the Login App

Reviewer 2: Can Login App example be linked/explained?
